### ISSUE TITLE
Skip auth form if valid refresh token exists?

### ISSUE THREAD

**Issue by madprime** [jazzband/django-oauth-toolkit#753]
According to [documentation on skipping the authorization form](https://django-oauth-toolkit.readthedocs.io/en/latest/advanced_topics.html#skip-auth-form), `auto` can be used to skip the authorization form if an application has already been authorized with the same scope.

However as it's currently implemented, this only seems to work as long as an access token exists that is valid (default expiration: 10 hours) and has matching scope. But it _doesn't_ check for a valid refresh token.

If a valid refresh token exists (and its corresponding access token has matching scopes), it seems the app has equivalent evidence of valid authorization from that user? And so one could (should?) safely skip the authorization form? Unless I'm missing something, this would seem to match the anticipated behavior as described in documentation.

If that all sounds correct, I think a change to `oauth2_provider/views/base.py` can make this happen, and I'll open a related PR. :)

**Comment by IvanAnishchuk** [comment]
Merged too quickly. Converting querysets to lists there is a bad idea.

**Comment by madprime** [comment]
I did expect more discussion – especially about the intended **change in logic made on the code for auto-authorization**, i.e. whether it is correct to change it in this way.

**Comment by IvanAnishchuk** [comment]
Yep. Let's continue the discussion then and figure out the correct way of doing it.

I think it is, generally speaking, a correct approach (it's common to skip authorization decision page if user has previously approved a request with the same scope for the same application, it also does make sense that refreshing a token or anything other than explicit user actions has side-effects on auto-authorization), but maybe we should look at the grants rather than/in addition to access/refresh tokens -- theoretically there can be access tokens issued for some other grant type that user didn't explicitly approve (not a common situation with DOT, applications support only a single grant type by default).

Existence of a Grant can confirm "user authorization decision" i.e. that user clicked the "authorize" button with the same scope and AccessToken or RefreshToken can confirm that user hasn't logged out (revoked access) since. I think we need to establish both facts to consider approval established "via other means" ([4.2.1](https://tools.ietf.org/html/rfc6749#section-4.2.1)).

There is, however, a slight problem with refresh tokens in DOT: they are not expiring. Unless you revoke tokens manually or change the code or run the management command that clears them they are sitting in the db forever (there is a settings variable that sets expiration period but it's never used except for the management command). Which means many times the authorization page will never reappear no matter what you do which is probably not always ideal. We could also try to use Grant timestamp for excluding very old tokens. Or something, the point here is establishing that the decision made by user to authorize the app is still valid, it's not always time based and not always action-based, perhaps the logic could vary from situation to situation or even be customizable like most important things in these libraries.

In short, we need to fix the following here: 1) excessive fetching of stuff from db, 2) check for refresh token expiration if the corresponding settings variable is set 3) check grants and not only tokens (DOT doesn't currently link them though so the exact logic needs further discussion)

**Comment by IvanAnishchuk** [comment]
Oh, forgot to mention: grants are also deleted by default when access token is issued.

**Comment by madprime** [comment]
@IvanAnishchuk I created a new PR (#755) that tries to avoid the queryset-to-list, and also check for refresh token expiration. It doesn't try to check grants (I figured this didn't need to block on that?).

BUT this is work in progress – it's failing tests and I got stuck. Maybe you or others could help? I made a comment here on what seemed to be the issue: https://github.com/jazzband/django-oauth-toolkit/pull/755/files#diff-7cdc0c1a73bed47e5369f87bb98b38e6R195

**Comment by dopry** [comment]
## Status: still valid, now blocked on ADR 0001 (#1723)

Keeping this open as a follow-up to #1723 rather than closing it — the request is legitimate, but it should not land as its own change. Recording where it stands.

### Current behaviour

`approval_prompt=auto` still scans **only live access tokens** (`oauth2_provider/authorization_server/views/base.py:329-347`):

```python
elif require_approval == "auto":
    tokens = (
        get_access_token_model()
        .objects.filter(user=request.user, application=kwargs["application"],
                        expires__gt=timezone.now())
        .all()
    )
    for token in tokens:
        if token.allow_scopes(scopes): ...
```

History for anyone arriving here: #754 implemented the refresh-token check and was merged 2019-10-24, then reverted in March 2020 following @IvanAnishchuk's review; the follow-up #755 stayed WIP and was closed in 2022. So master has never carried this.

### Two of the three 2019 objections have since been resolved

- **"Converting querysets to lists is a bad idea"** — cosmetic today, trivially avoidable.
- **"Refresh tokens never expire in DOT"** — ✅ **fixed.** `REFRESH_TOKEN_EXPIRE_SECONDS` is now enforced at *validation* time, not just by `cleartokens` (`oauth2_validators.py:1093-1110`, via #746), with the policy centralised in `refresh_token_expire_timedelta()` (`models.py:1461`). A liveness predicate now exists and is canonical, so an implementation would no longer have to invent one.
- **"Check Grants, not just tokens"** — ❌ **still infeasible.** The `Grant` row is deleted at code exchange (`oauth2_validators.py:462`), and there is no consent model. There is still no durable record that the user ever clicked *Authorize*.

Revocation is also in better shape than in 2019: `revoke_access_token()` (`models.py:1500`) is the single path shared by `/revoke/`, `AuthorizedTokenDeleteView` and the admin action, and it revokes through to the refresh token, so a user revoking an app would genuinely stop auto-approval.

### Why this is blocked rather than ready

The remaining hazard is that `REFRESH_TOKEN_EXPIRE_SECONDS` defaults to `None`. In the default configuration a live refresh token lives forever, so shipping the check as filed means: once a user consents, with `approval_prompt=auto` **the consent screen never appears again**. Gating it behind a new setting would work, but it deepens exactly the conflation ADR 0001 identifies as the root cause — *"the refresh token acting as a stand-in for both the user's consent and the user's session"* — and would add a setting that 4.0 then has to deprecate.

ADR 0001 (#1723, milestone 4.0) names this issue's problem directly under *"Consent is not a first-class fact"*:

> Whether a user has authorized an application is inferred from whether an unexpired access token happens to exist. Consent is forgotten when tokens expire and remembered only as long as tokens live.

Its Phase 1 scope includes *"optionally switch `require_approval="auto"` to consult `Authorization`"*, and **ADR open question 5 is precisely this decision** — Phase 1, or its own release behind a setting.

The `claude/adr-0001-phase1-authorization` branch already lands `Authorization` plus `authorization` FKs on `Grant`, `AccessToken`, `RefreshToken`, `IDToken` and `DeviceGrant`, and deliberately leaves the `require_approval == "auto"` block byte-identical to master. So the anchor this issue needs is being built; once question 5 is answered, this becomes a small diff against a table that already exists, instead of a policy invention in the view.

### Adjacent, tracked here so it isn't lost

`Authorization` currently records only `scope`. RFC 8707 `resource` is now carried on both token models and on the grant, and neither the present access-token scan nor any future consent check considers it — auto-approving a request for `resource=X` on consent recorded for `resource=Y` is wrong under either model. This is the same question the ADR already raises for RAR `authorization_details`, and is raised there rather than here.

---
_Generated by [Claude Code](https://claude.ai/code)_
